brainiak/matnorm_models/regression.py

```diff
@@ -0,0 +1,167 @@
+import tensorflow as tf
+import numpy as np
+from .helpers import define_scope
+from .base import MatnormModelBase
+
+
+class MatnormRegression(MatnormModelBase):
+    """ This analysis allows maximum likelihood estimation of regression models
+    in the presence of both spatial and temporal covariance.
+
+    ..math::
+    Y \\sim \\mathcal{MN}(X\beta, time_noise_cov, space_noise_cov)
+
+    Parameters
+    ----------
+    n_c : int
+        Number of columns in design matrix
+    n_v : int
+        number of voxels
+    time_noise_cov : subclass of NoiseCovBase
+        TR noise covariance class following NoiseCovBase interface.
+    space_noise_cov : subclass of NoiseCovBase
+        Voxel noise covariance class following NoiseCovBase interface.
+    learnRate : real, default=0.01
+        Step size for the Adam optimizer
+
+    """
+    def __init__(self, n_c, n_v, time_noise_cov, space_noise_cov,
+                 learnRate=0.01):
+
+        self.n_c, self.n_v, self.time_noise_cov, self.space_noise_cov =\
+            n_c, n_v, time_noise_cov, space_noise_cov
+
+        self.Y = tf.placeholder(tf.float64, [None, n_v], name="Y")
+
+        self.X = tf.placeholder(tf.float64, [None, n_c], name="X")
+
+        self.beta = tf.Variable(tf.random_normal([n_c, n_v], dtype=tf.float64),
+                                name="beta")
+
+        self.train_variables = [self.beta]
+
+        self.train_variables.extend(time_noise_cov.get_optimize_vars())
+        self.train_variables.extend(space_noise_cov.get_optimize_vars())
+
+        # create optimizer node
+        self.optimizer = tf.train.AdamOptimizer(learnRate)
+
+        # create a tf session we reuse for this object
+        self.sess = tf.Session()
+
+        # create training nodes
+        self.train_logp
+        self.train_optimize
+
+        self.sess.run(tf.global_variables_initializer())
+
+    @define_scope
+    def train_optimize(self):
+        """ Optimizer op (internal)
+        """
+        return self.optimizer.minimize(-self.train_logp,
+                                       var_list=self.train_variables)
+
+    @define_scope
+    def train_logp(self):
+        """ Log likelihood of model (internal)
+        """
+        y_hat = tf.matmul(self.X, self.beta)
+        resid = self.Y - y_hat
```

Comment:
`y_hat` is the estimate, yes, and `resid` is residual. I can rename both (though hat for estimate is pretty standard in stats I think).